### PR TITLE
Allow users to pass in sanitize method for XHR/fetch requests

### PR DESCRIPTION
**Pull request by SpennyNDaJets**
## Summary
While Highlight does provide whitelisting to [redact network header and body](https://www.highlight.io/docs/getting-started/client-sdk/replay-configuration/recording-network-requests-and-responses#redacting-headers-and-bodies), the current functionality is very limiting. It is a non-trival job to whitelist all sort of data types and keys for PIIs, tokens, passwords and so on by hand. It's specially harder when using graphql due to the single endpoint nature of the protocol.

Instead of having a rigid list of keywords and url rules, I'd like to be able to modify the entire network data to my liking based on my application needs and privacy rules. Allow for the client application to customize the request/response right before sending any data to its platform.

One can for instance, modify data from request/response headers, body etc as well as prevent the entire request/response to even be logged.

## How did you test this change?
1) Add in a `requestResponseSanitizer` to the `networkRecording` hash in the frontend's Highlight options. Example:
```
requestResponseSanitizer: (pair) => {
    if (pair.request.verb === 'POST') {
        console.log('blocking request', pair)
	return null
    } else {
	console.log('allowing request', pair)
	pair.request.headers ||= {}
	pair.request.headers['editTest'] = 'true'
    }

    return pair
}
```
2) Build the SDK and start a Highlight session
3) Click around to get different types of requests
4) Close the session, and let it process
5) View the processed sessions network logs
- [ ] Requests ignored appropriately
- [ ] Requests updated appropriately
- [ ] No known missing requests

https://www.loom.com/share/47ce07c349f14e9aa3fb802b21221167?sid=7153b018-f6ba-46a5-bff8-6c635b269607

## Are there any deployment considerations?
Monitor that Highlight records properly without the `requestResponseSanitizer` method

## Does this work require review from our design team?
N/A

### CODE REVIEW

**Comment by SpennyNDaJets** [comment]
@Vadman97 one call out that I wanted to make from what we discussed:

Opted that when a user returns `null` from this method, the request is not logged at all. I believe this is
1) More in line with the feature ask in https://github.com/highlight/highlight/issues/5948
2) Allows the user more flexibility - they can easily redact any data they feel fit, and should be allowed to ignore requests all together if that is their case.

File: docs-content/getting-started/3_client-sdk/7_replay-configuration/recording-network-requests-and-responses.md [hunk 1]
```diff
@@ -60,6 +60,38 @@ You can also redact specific headers by using `networkRecording.networkHeadersTo
 
 This configuration is only available for `highlight.run` versions newer than `4.1.0`.
 
+## Custom Sanitizing of Response and Requests
+
+If you need more granular control over managing and ignoring data in request and responses, then you can create a sanitize function to process every recorded request/response pair. This is configured by using the `networkRecording.requestResponseSanitizer` option.
+
+The `networkRecording.requestResponseSanitizer` method receives a RequestResponsePair, and should return an object of the same type or a `null` value. Returning a `null` value means that Highlight will drop the request, and no related network logs will be seen in the session replay.
```
Comment (deltaepsilon) [inline docs-content/getting-started/3_client-sdk/7_replay-configuration/recording-network-requests-and-responses.md:67]:
```suggestion
The `networkRecording.requestResponseSanitizer` method receives a Request/Response pair, and should return an object of the same type or a `null` value. Returning a `null` value means that Highlight will drop the request, and no related network logs will be seen in the session replay.
```

File: docs-content/getting-started/3_client-sdk/7_replay-configuration/recording-network-requests-and-responses.md [hunk 2]
```diff
@@ -60,6 +60,38 @@ You can also redact specific headers by using `networkRecording.networkHeadersTo
 
 This configuration is only available for `highlight.run` versions newer than `4.1.0`.
 
+## Custom Sanitizing of Response and Requests
+
+If you need more granular control over managing and ignoring data in request and responses, then you can create a sanitize function to process every recorded request/response pair. This is configured by using the `networkRecording.requestResponseSanitizer` option.
```
Comment (deltaepsilon) [inline docs-content/getting-started/3_client-sdk/7_replay-configuration/recording-network-requests-and-responses.md:65]:
```suggestion
Create a sanitize function to gain granular control of the data that your client sends to Highlight. The sanitize function is defined in the second argument of `H.init` under `networkRecording.requestResponseSanitizer`.
```

File: sdk/client/src/listeners/network-listener/utils/utils.ts [hunk 3]
```diff
@@ -31,10 +32,50 @@ type PerformanceResourceTimingWithRequestResponsePair =
...
+	) => RequestResponsePair | null
+}
+
+const sanitizeRequestResponsePair = (
+	pair: RequestResponsePair,
+	{
+		headersToRedact,
+		headersToRecord,
+		requestResponseSanitizer,
+	}: SanatizeOptions,
```
Comment (deltaepsilon) [inline sdk/client/src/listeners/network-listener/utils/utils.ts:49]: Spelling: `SanitizeOptions`

**Review by deltaepsilon — APPROVED** [review]

File: sdk/client/src/listeners/network-listener/utils/utils.ts [hunk 4]
```diff
@@ -108,26 +149,49 @@ export const matchPerformanceTimingsWithRequestResponsePair = (
...
+					performanceTiming.requestResponsePair
+
+				if (requestResponsePair) {
+					requestResponsePair = sanitizeRequestResponsePair(
+						performanceTiming.requestResponsePair,
+						sanitizeOptions,
+					)
+
+					// ignore request if it was filtered out by the user defined sanitizer
+					if (!requestResponsePair) {
```
Comment (Vadman97) [inline sdk/client/src/listeners/network-listener/utils/utils.ts:167]: cool, so returning `false`, `undefined,` or any other null-ish value would work

**Review by Vadman97 — APPROVED** [review]
looks good, but i'm struggling to get recording working - all network requests seem to be blank for config
```

		requestResponseSanitizer: (pair) => {
			if (Math.random() < 0.1) return null;
			return {
				...pair,
				response: {
					...pair.response,
					headers: {hello: 'world'},
				}
			}
		}
```
might be missing something, not sure
Screenshot 2023-10-25 at 8 21 11 PM

**Comment by SpennyNDaJets** [comment]
> d19676f

Good catch! I had a little regression in one of the refactors, but should be addressed now
Screenshot 2023-10-26 at 10 47 43 AM

**Pull request merged**